icloud-auth/src/client.rs

```diff
@@ -102,13 +101,59 @@ pub struct AppToken {
     pub app: String,
 }
 //Just make it return a custom enum, with LoggedIn(account: AppleAccount) or Needs2FA(FinishLoginDel: fn(i32) -> TFAResponse)
-pub enum LoginResponse {
-    LoggedIn(AppleAccount),
+#[repr(C)]
+#[derive(Debug)]
+pub enum LoginState {
+    LoggedIn,
     // NeedsSMS2FASent(Send2FAToDevices),
-    NeedsDevice2FA(),
-    Needs2FAVerification(),
-    NeedsLogin(),
-    Failed(Error),
+    NeedsDevice2FA,
+    Needs2FAVerification,
+    NeedsSMS2FA,
+    NeedsSMS2FAVerification(VerifyBody),
+    NeedsExtraStep(String),
+    NeedsLogin,
+}
+
+#[derive(Serialize, Debug, Clone)]
+struct VerifyCode {
+    code: String,
+}
+
+#[derive(Serialize, Debug, Clone)]
+struct PhoneNumber {
+    id: u32
+}
+
+#[derive(Serialize, Debug, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct VerifyBody {
+    phone_number: PhoneNumber,
+    mode: String,
+    security_code: Option<VerifyCode>
+}
+
+#[repr(C)]
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct TrustedPhoneNumber {
+    pub number_with_dial_code: String,
+    pub last_two_digits: String,
+    pub push_mode: String,
+    pub id: u32
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct AuthenticationExtras {
+    pub trusted_phone_numbers: Vec<TrustedPhoneNumber>,
+    pub recovery_url: Option<String>,
+    pub cant_use_phone_number_url: Option<String>,
+    pub dont_have_access_url: Option<String>,
+    pub recovery_web_url: Option<String>,
+    pub repair_phone_number_url: Option<String>,
+    pub repair_phone_number_web_url: Option<String>,
+    #[serde(skip)]
+    pub new_state: Option<LoginState>,
```

Comment:
why are they hardcoded? Couldn't they all be store in an hashmap?

Reply:
the log-in URL is also hardcoded, so I guess that's the reason

Reply:
They're hardcoded cuz serde is amazing :stuck_out_tongue:

Reply:
Yeah but imagine there is another second step we haven't anticipated, like idk `securityUpgrade`, having a hashmap (or the plist directly) of the urlbag would be more flexible

Reply:
It's also a different PR btw

Reply:
Plist has a `Value` type for this, repesents generic plist data, can be converted into structs and back